service/models/credit_card.py

```diff
@@ -0,0 +1,172 @@
+            self.card_number = data["card_number"]
+            self.expiry_month = data["expiry_month"]
+            self.expiry_year = data["expiry_year"]
+            self.security_code = data["security_code"]
+            self.billing_address = data["billing_address"]
+            self.zip_code = data["zip_code"]
+        except AttributeError as error:
+            raise DataValidationError("Invalid attribute: " + error.args[0]) from error
+        except KeyError as error:
+            raise DataValidationError(
+                "Invalid CreditCard: missing " + error.args[0]
+            ) from error
+        except TypeError as error:
+            raise DataValidationError(
+                "Invalid CreditCard: body of request contained bad or no data "
+                + str(error)
+            ) from error
+        return self
+
+    ##################################################
+    # VALIDATIONS
+    ##################################################
+
+    @validates("first_name")
+    def validate_first_name(self, _key, first_name):
+        """Validates `first_name` field"""
+        if is_not_str(first_name) or not first_name.isalpha():
+            raise DataValidationError("First name field must contain letters only")
+
+        return first_name
+
+    @validates("last_name")
+    def validate_last_name(self, _key, last_name):
+        """Validates `last_name` field"""
+        if is_not_str(last_name) or not last_name.isalpha():
+            raise DataValidationError("Last name field must contain letters only")
+
+        return last_name
+
+    @validates("card_number")
+    def validate_card_number(self, _key, card_number):
+        """Validates `card_number` field"""
+        if is_not_str(card_number) or not card_number.isdigit():
+            raise DataValidationError("Card number field must be numeric")
+
+        if not len(card_number) == 16:
+            raise DataValidationError("Card number field must be 16 digits")
+
+        return card_number
+
+    @validates("security_code")
+    def validate_security_code(self, _key, security_code):
+        """Validates `security_code` field"""
+        if is_not_str(security_code) or not security_code.isdigit():
+            raise DataValidationError("Security code field must be numeric")
+
+        if not len(security_code) == 3:
+            raise DataValidationError("Security code field must be 3 digits")
+
+        return security_code
+
+    @validates("expiry_year")
+    def validate_expiry_year(self, _key, expiry_year):
+        """Validates `expiry_year` field"""
+        if is_not_int(expiry_year) or expiry_year < 2024 or expiry_year > 2050:
+            raise DataValidationError("Expiry year field is invalid")
+
+        return expiry_year
+
+    @validates("expiry_month")
+    def validate_expiry_month(self, _key, expiry_month):
+        """Validates `expiry_month` field"""
+        if is_not_int(expiry_month) or expiry_month < 1 or expiry_month > 12:
```

Comment:
I think we need to combine the expiry month and expiry year, since if the expiry date is 02 2024, now it has already expired but it is still available here

Reply:
I think it's okay to be lenient with expiry date validation since the date will expire to any card anyways. it's usually handled with some sort of warning, since the card is valid but it just wont be processed.

Reply:
That's ok, I got your point